### PR TITLE
[CF-237] Migrate detached ports.

### PR DESCRIPTION
**Pull request by antonf**
This patch add migration of detached ports (not attached to VM or
router) to network deployment procedure.

### CODE REVIEW

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2665/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.utils.cache
cloudferrylib/utils/cache.py:15: [W0611(unused-import), ] Unused import collections
************* Module cloudferrylib.utils.log
cloudferrylib/utils/log.py:20: [W0611(unused-import), ] Unused import sys
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2666/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.utils.cache
cloudferrylib/utils/cache.py:15: [W0611(unused-import), ] Unused import collections
************* Module cloudferrylib.utils.log
cloudferrylib/utils/log.py:20: [W0611(unused-import), ] Unused import sys
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

File: cloudferrylib/utils/log.py [hunk 1]
```diff
@@ -61,7 +61,7 @@ def configure_logging():
         for handler in logger.handlers:
             if handler.name == 'console':
                 handler.setLevel(logging.DEBUG)
-    sys.stdout = StdoutLogger()
+    #sys.stdout = StdoutLogger()
```
Comment (roman-verchikov) [inline cloudferrylib/utils/log.py]: Uncomment or move to config

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.utils.cache
cloudferrylib/utils/cache.py:15: [W0611(unused-import), ] Unused import collections
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2674/

**Comment by roman-verchikov** [comment]
Please add description to pull request and commit message

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check failed
http://jenkins-cf.vm.mirantis.net:8080/job/cloudferry-unittests-and-style-checks/2679/

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.utils.cache
cloudferrylib/utils/cache.py:15: [W0611(unused-import), ] Unused import collections
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by antonf** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), NeutronNetwork.create_port] Arguments number differs from overridden 'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), NeutronNetwork.check_existing_port] Arguments number differs from overridden 'check_existing_port' method
cloudferrylib/os/network/neutron.py:833: [W0221(arguments-differ), NeutronNetwork.get_security_groups] Arguments number differs from overridden 'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/network/neutron.py [hunk 2]
```diff
@@ -81,6 +81,9 @@ def read_info(self, **kwargs):
 
         nets = self.get_networks(tenant_id)
         subnets = self.get_subnets(tenant_id)
+        detached_ports = self.get_detached_ports(tenant_id)
+        LOG.debug('List of detached ports: %s',
```
Comment (miarmak) [inline cloudferrylib/os/network/neutron.py]: Maybe it is better to have this log message inside the `get_detached_ports` method?

File: cloudferrylib/os/network/neutron.py [hunk 3]
```diff
@@ -1232,7 +1251,41 @@ def upload_networks(self, networks, src_seg_ids):
...
+                net['id'], port['mac_address'],
+                ip_addresses=ip_addresses,
+                existing_ports=existing_ports.values())
+            if existing_port is not None:
+                if existing_port['mac_address'] == port['mac_address']:
+                    LOG.debug('Port %s already migrated to %s',
+                              port['id'], existing_port['id'])
+                    continue
+
+                if existing_port['device_owner'].startswith('network:') or \
```
Comment (miarmak) [inline cloudferrylib/os/network/neutron.py]: Have you taken into account device owners, that start with `neutron:`?

**Comment by roman-verchikov** [comment]
Please check grizzly to juno migration log, for some reason VM was inaccessible over floating IP, might be somehow related:
```
RuntimeError: VM with ip "192.168.1.107" is not accessible via ssh after 20 attempts
```

**Comment by SVilgelm** [comment]
Please take a look at grizzly to juno log.
The error:
```
IpAddressInUseClient: Unable to complete operation for network d8fa62be-f7f7-4e79-a3b2-40899e572e68. The IP address 10.4.2.2 is in use.
```

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:835: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by antonf** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:835: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:835: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

File: cloudferrylib/os/network/neutron.py (hunk 2 above)
Comment (antonf) [inline cloudferrylib/os/network/neutron.py]: nope

**Comment by antonf** [comment]
Found why
`IpAddressInUseClient: Unable to complete operation for network d8fa62be-f7f7-4e79-a3b2-40899e572e68. The IP address 10.4.2.2 is in use.`
happened: dictionary returned by `create_network` didn't have subnets, so DHCP wasn't disabled for them. Fixed it.

**Comment by antonf** [comment]
retest this please

**Comment by MirantisWorkloadMobility** [comment]
Consider resolving pylint issues: \
``` \
************* Module cloudferrylib.os.network.neutron
cloudferrylib/os/network/neutron.py:188: [W0221(arguments-differ), 
NeutronNetwork.deploy] Arguments number differs from overridden 'deploy' method
cloudferrylib/os/network/neutron.py:294: [W0221(arguments-differ), 
NeutronNetwork.create_port] Arguments number differs from overridden 
'create_port' method
cloudferrylib/os/network/neutron.py:347: [W0221(arguments-differ), 
NeutronNetwork.check_existing_port] Arguments number differs from overridden 
'check_existing_port' method
cloudferrylib/os/network/neutron.py:835: [W0221(arguments-differ), 
NeutronNetwork.get_security_groups] Arguments number differs from overridden 
'get_security_groups' method

**Comment by MirantisWorkloadMobility** [comment]
PEP8 check success, Unittests success

**Comment by roman-verchikov** [comment]
Looks good to me, but requires testing in staging

**Comment by wzbbbb** [comment]
+1
tested in DPA1

**Comment by roman-verchikov** [comment]
can be merged

**Pull request merged**